### DISCUSSION TITLE
What is the Use Case for Using `*args` as a Type Variable Tuple?

### DISCUSSION THREAD

**Discussion by adam-grant-hendry** [python/typing#1258] [Q&A]
I don't understand the intended use case for [`*args` as a Type Variable Tuple](https://peps.python.org/pep-0646/#args-as-a-type-variable-tuple) as laid out in `PEP 646`.

I don't understand why I wouldn't do [what's recommended by typeshed](https://typing.readthedocs.io/en/latest/source/best_practices.html#using-any-and-object) and give `*args` the type `object` so it can be any type and then type narrow in the function body. It seems like the target method that processes `*args` is going to have to type narrow any way if its items are truly indeterminate in size and type.

Maybe it is used for top-level functions that use helper functions that accept `*args`? The `PEP` provides `os.execle` as an example, which [passes `args` to `os.execve`](https://github.com/python/cpython/blob/78359b1d45608439f8e03b8e86174fe7b04d3e08/Lib/os.py#L543), but `typeshed` seems to [manage with using `Any`](https://github.com/python/typeshed/blob/20bc8fa4e30df5f879606dfc37dd0b4bcebaeedd/stdlib/os/__init__.pyi#L795) (granted, `TypeVarTuple` isn't supported by `mypy` yet). Again though, I would have just typed this as `object`.

What is the intended use case for [`*args` as a Type Variable Tuple](https://peps.python.org/pep-0646/#args-as-a-type-variable-tuple)?

#### Update

Please feel free to also review [my SO answer](https://stackoverflow.com/questions/37031928/type-annotations-for-args-and-kwargs/73738408#73738408) related to typing `*args` and `**kwargs`. I do not know when one should use `TypeVarTuple` for `*args`.

**Comment by TeamSpen210** [comment]
One use case is to pair it with a `Callable`, where the `*args` are passed in when it's called (perhaps with additional parameters added before or after). In that situation it serves a similar purpose to `ParamSpec`, except where keyword arguments aren't allowed. 

Though you might need to be type narrowing inside the function, the important thing is that it lets you then transfer those types to the return type in whatever way is appropriate.

**Reply by erictraut** [reply]
The main use is for generic classes with variadic type variables. 

Consider the following:
```python
Ts = TypeVarTuple("Ts")

Width = NewType("Width", int)
Height = NewType("Height", int)
Depth = NewType("Depth", int)

class Vector(Generic[*Ts]):
    def __init__(self, *args: *Ts) -> None:
        ...
    
    def add(self, *args: *Ts) -> "Vector[*Ts]":
        ...

    def as_tuple(self) -> tuple[*Ts]:
        ...

v3 = Vector(Width(4), Height(3), Depth(1))
reveal_type(v3) # Vector[Width, Height, Depth]

v2 = Vector(Width(0), Height(0))
reveal_type(v2) # Vector[Width, Height]
v2.add(Width(1), Height(5))

reveal_type(v2.as_tuple()) # tuple[Width, Height]
```

**Reply by adam-grant-hendry** [reply]
@erictraut Your use case makes the most sense. Why do none of the examples in [`*args` as a Type Variable Tuple](https://peps.python.org/pep-0646/#args-as-a-type-variable-tuple) express this under the context of NumPy and TensorFlow arrays? The primary motivation for `PEP 646` was to support typing such data structures.

The [`*args` as a Type Variable Tuple](https://peps.python.org/pep-0646/#args-as-a-type-variable-tuple) section is extremely misleading. It begins with:

> [PEP 484](https://peps.python.org/pep-0484) states that when a type annotation is provided for `*args`, every argument must be of the type annotated. That is, if we specify `*args` to be type `int`, then all arguments must be of type int. This limits our ability to specify the type signatures of functions that take heterogeneous argument types.

The reader who reads this naturally thinks the author is claiming to solve the existing `*args` limitation _for all cases_, particularly in the general sense. I claim you are going to get a lot of new users who read that section and will use it incorrectly. I have at least [one example](https://stackoverflow.com/questions/37031928/type-annotations-for-args-and-kwargs/73738408#73738408), but I'm sure the question has been asked many times on SO. And as we discussed in the [pyright discussions](https://github.com/microsoft/pyright/discussions/3953#discussioncomment-3658712), the wording promotes bad practices and not using `TypeVar`s appropriately.

I would never have read this section of the PEP and known that the above was the intended use case.

I strongly believe this section needs to be removed and rewritten. It's going to cause a lot of misuse, misunderstanding, and misinformation in the python community.

**Reply by adam-grant-hendry** [reply]
> One use case is to pair it with a Callable, where the *args are passed in when it's called (perhaps with additional parameters added before or after). In that situation it serves a similar purpose to ParamSpec, except where keyword arguments aren't allowed.
> 
> Though you might need to be type narrowing inside the function, the important thing is that it lets you then transfer those types to the return type in whatever way is appropriate.

@TeamSpen210 If `ParamSpec` can do the same thing, I doubt [`*args` as a Type Variable Tuple](`https://peps.python.org/pep-0646/#args-as-a-type-variable-tuple) would have been added to `PEP 646`.

A requester must require showing proof of need without undue burden to the python language and/or typing ecosystem before a `PEP` can even be put into the running for being accepted. I can see the ability to use [`*args` as a Type Variable Tuple](`https://peps.python.org/pep-0646/#args-as-a-type-variable-tuple) as a side effect of adding tuple unpacking and `TypeVarTuple` to the ecosystem, but not as a feature. The way the section is worded in the `PEP` indicates it as an added feature to the typing ecosystem.

That said, your suggestion seems aligned with what's written in the motivation section of `PEP 692`. Although the explanation is w.r.t to `**kwargs`, the same principles apply to `*args`:

> Moreover, `**kwargs` can be used to reduce the amount of code needed in cases when there is a top-level function that is a part of a public API and it calls a bunch of helper functions, all of which expect the same keyword arguments. Unfortunately, if those helper functions were to use `**kwargs`, there is no way to properly type hint them if the keyword arguments they expect are of different types.

However, I don't think what you're suggesting was intended as a use case.

@erictraut I think the above speaks even further to confusion this section of `PEP 646` causes.

**Reply by erictraut** [reply]
I'll note that a PEP is a point-in-time specification for a proposed new piece of functionality. PEPs don't represent the latest (or best) documentation for a feature. On rare occasions, PEPs are amended in small ways to eliminate confusion, but this involves small (typically one-line) changes.

The feature described in PEP 646 is a very advanced typing feature. To my knowledge, no mainstream libraries have yet to adopt it, and that's unlikely to change until mypy implements it. It is fully supported by pyright and pyre, but I don't think mypy, pytype or PyCharm have any support for it yet. I personally implemented it in pyright, so I can attest that this is a very complex feature to implement. It may be a long time before it is implemented everywhere. That doesn't mean that it's a bad feature. I think the PEP does a good job explaining the motivation. There is a widely-recognized need for better typing support for tensors. I think you may be focusing on one aspect of PEP 646 (`*args` type annotations) and attempting to use it in a manner for which it wasn't designed. If you go back and read the full PEP, including the motivation section at the top, it might become clearer.

**Reply by adam-grant-hendry** [reply]
> That doesn't mean that it's a bad feature. 

I never said, nor suggested, that this is a bad feature. `TypeVarTuple` is wonderful. I only said I don’t understand the `*args` usage.

> I think the PEP does a good job explaining the motivation.

With respect to the `*args` section, I disagree. PEPs are written, implemented, and approved by advanced users. Consequently, you run a greater risk of not explaining things well because it’s harder for you to think like a new user. It making sense to you and possibly to advanced users does not mean that it has been explained well.

> I think you may be focusing on one aspect of PEP 646 (*args type annotations) and attempting to use it in a manner for which it wasn't designed.

The issue is that it’s not just me. I have shown that many others misunderstand what is meant by this section. You can see it in the issue tracker and SO.

Even if a PEP represents a point in time, it lives for quite a while and becomes our guiding document. This is not just typing PEPs, but all PEPs.

Users should be encouraged to read, understand, and utilize PEPs. Simply because something is deemed advanced does not mean it belongs only to advanced users.

**Reply by adam-grant-hendry** [reply]
@erictraut Note also that [in your example](https://github.com/python/typing/discussions/1258#discussioncomment-3672115), the methods `__init__()` and `as_tuple()` have `Ts` appear only once, but [this works](https://pyright-playground.decorator-factory.su/?gzip=H4sIADhIJ2MC_32QP2-DMBDFd3-KE13siqIktEukdqrUTploOiBkITgSq8Qg202bb19jQ_gjpR4s3z2_n-5epZoTmEsr5AHEqW2UgTeUqERBqlHi-GtQatFIPfza4U9yaTGE7t7nKvlua1t9yDYvvghJNDzPJBokOmCEfIrSHK3W-2ngGkEIQhpG3lEcjmYq-86gv2I7d7vGoJKizrWGPRamUbTfI_UjpYnOMrYlYE-JFXAupDCcU411FcJ9rg56C-NfBg8vsGskekt3oihy7ysjL8t_7YGfZDpBMMeNKM2Ni6njObMrp9aF8xzbHPpVXYb0kYXg46Kxfbpo6JoxovCMec1Nl9g5ZnDX-1LnG0y9I7PozRK9GtGrJXBzC5hZUNRl5BnrkfFkGQtIdE2AdTy__QL3B0R_SXetAgAA) because its contained within the context of a `Generic`, and that usage isn't made clear within the `PEP`.

From [your explanation](https://github.com/microsoft/pyright/discussions/3953#discussioncomment-3658712), it makes perfect sense that a type variable by itself, outside of the context of mapping between generic types, serves little (or no) value.

Could these usage notes be added to one or both of these?:

- [Proposal: Convert the typing-module PEPs into a series of HOWTOs](https://github.com/python/cpython/issues/91533)
- [Document Structure and Wishlist](https://github.com/python/typing/issues/845)

**Reply by AlexWaygood** [reply]
> From [your explanation](https://github.com/microsoft/pyright/discussions/3953#discussioncomment-3658712), it makes perfect sense that a type variable by itself, outside of the context of mapping between generic types, serves little (or no) value.

This isn't correct. We tried to elaborate some of the other use cases in the docs for the Python 3.11 `typing` module, which we tried to make slightly more accessible than the PEP (since the PEP, as Eric describes, isn't meant as end-user documentation, and isn't meant to be a living document): https://docs.python.org/3.11/library/typing.html#typing.TypeVarTuple

> Could these usage notes be added to one or both of these?:

Information on `TypeVarTuple`s and variatic generics would be included in part 2.vi of https://github.com/python/cpython/issues/91533, if I (or anybody else) ever gets round to putting significant work into actualising that proposal :)

**Reply by adam-grant-hendry** [reply]
> This isn't correct

Just when I thought I couldn't get more confused...

Which part? My understanding or @erictraut 's explanation? [His claim](https://github.com/python/typing/discussions/1258#discussioncomment-3672115) is the primary use case for typing `*args` with a `TypeVarTuple` is when used within `Generic`s. This seems to contradict the [3.11 docs](https://docs.python.org/3.11/library/typing.html#typing.TypeVarTuple).

If there are use cases beyond the `Generic`, it seems the `pyright` warning

> TypeVar "Ts" appears only once in generic function signature

is useful as an opt-in feature for a code reviewer to quickly find the locations where this occurs, but might lead to a significant number of false-positives, and that the statement

>  it's generally indicative of a user's misunderstanding of how type variables work. If the use of a TypeVar is unnecessary and provides no value in a particular case, you are probably doing something wrong.

is not correct.

Secondly, is `TypeVarTuple` meant to handle the case of `*args` being a heterogenous type of indeterminate arity? That is, is it meant to handle the case of a top-level function calling multiple helpers and preserving type information between calls, similar to `ParamSpec` and what @TeamSpen210 [suggested](https://github.com/python/typing/discussions/1258#discussioncomment-3671928)?

**Reply by AlexWaygood** [reply]
Eric is of course correct that the usage of `TypeVarTuple` in generic classes is the primary use case the PEP authors had in mind, and this is probably going to be its primary use case in the world of scientific computing. But I think you're incorrect to say that other uses of `TypeVarTuple` have "little (or no) value". Please take a look at the examples we added to the typing documentation for 3.11 to see some of the other use cases: https://docs.python.org/3.11/library/typing.html#typing.TypeVarTuple.

I've also listed some of the places where typeshed might be able to use `TypeVarTuple` to improve its stubs here: https://github.com/python/typeshed/issues/8708#issuecomment-1241844607. Note that only one of these use cases involves a generic class parameterised by a `TypeVarTuple`; the majority of use cases involve generic functions. (This is mostly because typeshed does not have very many scientific-computing stubs packages in it, but is still illustrative of the fact that there are other use cases for `TypeVarTuple` outside of generic classes.)

**Reply by adam-grant-hendry** [reply]
> But I think you're incorrect to say that other uses of `TypeVarTuple` have "little (or no) value".

"little (or no) value" does sound excessive, I agree. I was thinking I shouldn't use it for top-level functions calling helpers to preserve type information between calls, but this [does seem to be a valid use case](https://docs.python.org/3.11/library/typing.html#typing.TypeVarTuple). Without that, I couldn't think of a common use case for it besides what Eric mentioned, which is why I said what I said.

Thank you for showing me the additional use cases! I'll consider Eric's use case in addition to the [python docs use cases](https://docs.python.org/3.11/library/typing.html#typing.TypeVarTuple).
